apps/shipments/models.py

```diff
@@ -93,6 +98,31 @@ def geocoder(self, parsing_address, method):
             self.geometry = Point(geocoder_response.latlng)
 
 
+class Device(models.Model):
+    id = models.CharField(primary_key=True, null=False, max_length=36)
+    certificate_id = models.CharField(unique=True, null=True, blank=False, max_length=255)
+
+    @staticmethod
+    def get_or_create_with_permission(jwt, device_id):
+        certificate_id = None
+        if settings.PROFILES_URL:
+            # Make a request to Profiles /api/v1/devices/{device_id} with the user's JWT
+            response = requests.get(f'{settings.PROFILES_URL}/api/v1/device/{device_id}/',
+                                    headers={'Authorization': 'JWT {}'.format(jwt.decode())})
+            if response.status_code != status.HTTP_200_OK:
+                raise PermissionDenied("User does not have access to this device in ShipChain Profiles")
+
+        if settings.ENVIRONMENT != 'LOCAL':
+            import boto3
+            iot = boto3.client('iot')
+            response = iot.list_thing_principals(device_id)
+            for arn in response['principals']:
```

Comment:
From what I understand, a 'Principal' is an association between a device and a certificate in IoT. Devices have certificates provisioned for them when they are created in IoT. Just as a test I created one without a device - there's no exception thrown by boto3 but the list of principals is empty. This would create a `Device` in transmission without any certificate_id, which would result in 403 errors if that device tried to do a tracking update.

Good catch, I think it's probably a good idea to enforce that there be a certificate attached to a device. Also needed to handle the case in which a device is not found in IoT. :+1: